Post-Encryption Followup

tcaputi opened the pull request:
### Description
This patch includes 2 main pieces: fixes for bugs and documentation issues found after the encryption patch (b52563) was merged and general code improvements for long-term maintainability. Integration of encryption with the new `zfs recv -o / -x` capabilities will be moved to a separate PR after this one is merged

### Motivation and Context

##### Bug Fixes and Feature additions

- Most of the bugs and documentation issues are from #6547. (Thanks to @stewartadam for finding most of them).
- Another bug found by @stewartadam regarding raw properties sends of non-encryption roots (#6624).
- Several problems and race conditions described in #6639 
- Key changing bug from #6703 
- Send / Receive bug from #6706 

##### Maintainability Improvements (no functional changes)

- `ztest` now creates and properly tests encrypted datasets.
- PBKDF2 implementation changed to use libssl instead of custom implementation
- HKDF moved to its own file and tests were added to `zfs-test` to ensure its correctness
- Removed now unnecessary dependency of userspace libraries on libzpool and libicp
- Updated man pages
- Fixed Coverity defects



### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

tcaputi commented:
Other than ensuring that buildbot is happy with this there are 2 outstanding pieces:

- I am currently working on a zfs-test for the new `zfs recv -o / -x` integration.
- There is a problem with `zfs recv -o encryption=...` that I had not anticipated. I only thought of it when I went to write a quick test script after the implementation was complete. The problem is that `zfs recv` currently expects the send stream to be piped in via stdin. Normally this is fine, but the binary also needs to be able to prompt the user for a key / passphrase if he / she selects `keylocation=prompt`. I'm not quite sure how to solve this issue but I hope to come up with something soon....

behlendorf commented on man/man8/zfs.8:
```diff
@@ -3445,6 +3453,15 @@ is a special case because, even if
+.Nm zfs Cm send Fl w
+) may only be received as is, and cannot be reencrypted, decrypted, or
+recompressed by the receive process. Unencrypted streams can be received as
+encrypted datasets, either through inheritence or by specifying encryption
```
s/inheritence/inheritance

behlendorf commented on man/man8/zfs.8:
```diff
@@ -3445,6 +3453,15 @@ is a special case because, even if
 .Pp
+Raw encrypted send streams (created with
+.Nm zfs Cm send Fl w
+) may only be received as is, and cannot be reencrypted, decrypted, or
```
s/reencrypted/re-encrypted

behlendorf commented on module/zfs/dmu_send.c:
```diff
@@ -542,6 +542,8 @@ dump_dnode(dmu_sendarg_t *dsp, const blkptr_t *bp, uint64_t object,
 		drro->drr_blksz = SPA_OLD_MAXBLOCKSIZE;
 
+	bonuslen = P2ROUNDUP(dnp->dn_bonuslen, 8);
+
```
Down a few lines, s/the the/the

behlendorf commented on module/zfs/hkdf.c:
```diff
@@ -0,0 +1,192 @@
+	ret = crypto_mac(&mech, &input_cd, &key, NULL, &output_cd, NULL);
+	if (ret != CRYPTO_SUCCESS) {
+		ret = SET_ERROR(EIO);
+		goto error;
```
Just `return (SET_ERROR(EIO));`.  The same goes for the `goto`'s in`hkdf_sha512_expand()` and `hkdf_sha512`.  As a general rule if there's no state in the function which needs to be unwound you don't need to return through a `goto`.

behlendorf commented on rpm/generic/zfs.spec.in:
```diff
@@ -87,6 +87,7 @@ BuildRequires:  libuuid-devel
 BuildRequires:  libblkid-devel
 BuildRequires:  libudev-devel
 BuildRequires:  libattr-devel
+BuildRequires:  openssl-devel
```
The core zfs utilities now require `openssl` so we need to add:

```
Requires: openssl
```

I believe dracut will automatically pull in the library dependency for the initramfs.  However, while looking at `./contrib/dracut/90zfs/module-setup.sh.in` a noticed that `installkernel()` has never included the icp module and it should make sure that gets installed.

behlendorf commented on tests/zfs-tests/tests/functional/hkdf/.gitignore:
```diff
@@ -0,0 +1 @@
+hkdf_test
```
Most of the new files are missing a newline at the end of file.  Please add it.

behlendorf commented on tests/zfs-tests/tests/functional/hkdf/hkdf_test.c:
```diff
@@ -0,0 +1,234 @@
+/*
+ * XXX Neither NIST nor IETF has published official test
+ * vectors for testing HKDF with SHA512. The following
+ * vectors should be updated if these are ever published.
```
Since we didn't get these from NIST of IETF it's worth a comment explaining where they did come from.

behlendorf commented on tests/zfs-tests/tests/functional/hkdf/hkdf_test.c:
```diff
@@ -0,0 +1,234 @@
+
+	printf("Test failed.\n");
+	return (ret);
+}
```
Instead how about this.  Note that we always want to make sure we return `1` on failure.  Other non-zero values could result in problems when scripting if those values aren't well defined.

```c
int
main(int argc, char **argv)
{
	int error = 0;

        icp_init();

	for (int i = 0; i < NELEMS(test_vectors); i++) {
		error = run_test(i, &test_vectors[i]);
		if (error != 0)
			break;
	}

	icp_fini();

        if (error == 0) {
            	printf("All tests passed successfully.\n");
                return (0);
        } else {
        	printf("Test failed.\n");
                return (1);
        }
}
```

behlendorf commented on tests/zfs-tests/tests/functional/hkdf/run_hkdf_test.ksh:
```diff
@@ -0,0 +1,30 @@
+
+log_assert "Run the tests for the HKDF algorithm."
+
+log_must $STF_SUITE/tests/functional/hkdf/hkdf_test
```
If you add `hkdf_test` to `ZFSTEST_FILES` in `tests/zfs-tests/include/commands.cfg` a link to it will be created in the constrained path and you can rewrite this as:

```
log_must hkdf_test
```

nwf commented:
@tcaputi The typical thing to do when stdin is redirected is to open /dev/tty and do your prompting and read via that fd.  That's how, e.g. ssh does its thing.

tcaputi commented:
@nwf What exactly do you mean by that? Could you provide an example of how this would look on the command line UI?

nwf commented:
@tcaputi https://github.com/openssh/openssh-portable/blob/HEAD/openbsd-compat/readpassphrase.c shows off the technique (under the other cruft).  The key bit being

```
	if ((flags & RPP_STDIN) ||
	    (input = output = open(_PATH_TTY, O_RDWR)) == -1) {
		if (flags & RPP_REQUIRE_TTY) {
			errno = ENOTTY;
			return(NULL);
		}
		input = STDIN_FILENO;
		output = STDERR_FILENO;
	}
```

which, unless instructed otherwise, opens /dev/tty for interaction with the user.  Even if stdin has been redirected, this will still attach to the controlling terminal (assuming there is one).  The rest of that function is largely an effort to work around the UNIX TTY layer being pretty gross.  (This notion of "controlling terminal" is why paranoid applications open everything `O_NOCTTY`, just in case the thing they open turn out to be a terminal, which they might otherwise implicitly adopt as their controlling terminal.  It's a mess.)

What it looks like from the command line is pretty standard:
```
pf% ssh gradx.cs.jhu.edu cat < tmp/foo > tmp/bar 2> tmp/baz
nwf@gradx.cs.jhu.edu's password: 
```

nwf commented:
@tcaputi If you don't want to play that game, tho', the other popular option is to add a command line argument to allow passphrases to be read in from a file or some other source (see openssl's "Pass Phrase Options" at e.g. https://www.openssl.org/docs/manmaster/man1/openssl.html for some typical options) or to pawn off the job to some other program, a la openssh's `SSH_ASKPASS` environment variable (which is tied to `DISPLAY` but need not be, in general).

tcaputi commented:
@nwf Those are interesting solutions. I will leave it up to @ahrens to decide which one is best.

The only problem I have with the first solution is that (if I'm understanding it right) it is inherently very hard to test via scripts since the password can't be given without user interaction.

The second option basically describes my current workaround for testing which is using `keylocation=file://...` to ensure a prompt is never needed. I could easily add a check for this case to make sure the user can't ask for a `prompt` passphrase at recv time. The user could then manually change the location just with `zfs set keylocation=prompt`. But this doesn't seem ideal.

MyPod commented:
@tcaputi I'd think that `expect` could be used to make the first solution automated in tests, but I'm not familiar enough with it to suggest a sample script.

lundman commented:
and possibly `script`.

sylveon commented:
Why not both? A way for the user to specify passphrase on the prompt and a
way for them to specify location of a passphrase file in the command line.

One could also use the pinentry utility to prompt the user for a passphrase
via a graphical dialog when available.

On sam. 2 sept. 2017 07:43 Jorgen Lundman <notifications@github.com> wrote:

> and possibly script.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zfsonlinux/zfs/pull/6595#issuecomment-326739246>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGJFthV5FZk4Aod4rpQqaBVAyfy62zUBks5seT9_gaJpZM4PKjdg>
> .
>

sempervictus commented:
Expect isn't always around, found out the hard way when writing TTY shells for Metasploit. Ended up settling on a dual approach with expect and the corresponding python module, and still don't have great coverage. Since this is all in C anyway, the relevant functionality may be beneficial within the userspace utils themselves - the fewer deps the better IMO, especially given that this has to run on Linux, Illumos, OSX, and BSD (at least).

tcaputi commented:
At this point we should be getting pretty close to ready. @behlendorf I have addressed all of your comments when you are ready to check over them again. I apologize for the delay; enabling ztest on encrypted datasets revealed a few minor bugs which needed to be tracked down. I will monitor buildbot as tests come in and make sure ztest is happy.

tcaputi commented:
Well... it looks like I spoke too soon. I'll keep at it then.

lundman commented:
Out of curiosity, what was the deciding factor to link against OpenSSL instead? Probably does not affect Linux much, beyond that you would need OpenSSL in your initramfs. 

I'm not sure if OpenSSL is even part of IllumOS build so I'm unsure what can be done there.

tcaputi commented:
@lundman License compatibility pretty much. Needs to work with both GPL and CDDL.

nwf commented:
Not to ask an ignorant question, but could the tools statically link the PBKDF functions from OpenSSL, or could they be statically linked into libz*?  That way we'd not need all of libcrypto.so in the initramfs?  Not that it matters much to me, but I could see some people balking at the size.

tcaputi commented:
@nwf Thats definitely a possibility. If enough people care (and I'm not sure anyone does) we can look into doing something like that.

behlendorf commented:
@lundman additionally we wanted to depend on a known good and trusted implementation of the pbkdf2 function.  This is unfortunately one of those a places where each platform should really do the right thing for their system.

> could the tools statically link the PBKDF functions from OpenSSL,

That's a good idea.  We use such a tiny bit of OpenSSL it would be nice to statically link in only what we need in order to keep the initramfs as small as possible.  Plus keeping the dependencies to a minimum as always good.

tcaputi commented:
Submitted a bunch of fixes for things found since last week. Unfortunately, I had to disable ztesting for encrypted datasets since there are one or two unrelated bugs which come to light under the encryption code paths. These will be re-enabled as soon as they are fixed.

I will watch buildbot again to see what happens.

tcaputi commented:
As a side note, I looked into statically linking against libopenssl-devel. Unfortunately, it seems this will not be possible from a license compatibility standpoint. However, the libopenssl-devel shared objects only seem to be < 2MB in size so it shouldn't be too big of a deal to keep these in the initramfs.

sempervictus commented:
Any chance libressl would work here? Smaller by a longshot, and I believe has more BSD-friendly licensing.

behlendorf commented:
Maybe, but we'd like to choose something which is commonly available for all the Linux distributions and in the package manages.  I'm not sure if that's true of LibreSSL.  Plus wasn't it forked from the OpenSSL and therefore bound by the same license?

tcaputi commented:
@sempervictus It still wouldn't work statically. It has the same license as libopenssl-devel: https://github.com/libressl/libressl/blob/master/src/LICENSE

I think we would prefer to use openssl here though, simply because it is much more ubiquitous on Linux.

tcaputi commented:
Looks like somehow this patch messes with pool creation and freezing? I don't quite understand whats going on here, but I'll try to figure it out tomorrow morning.

sempervictus commented:
@tcaputi: still running into the problem of trying to send from unencrypted datasets to encrypted trees via _zfs send -Rve_ - complains that it can't recv the unencrypted property. Recv side propertly (-o -x) not helping here. The target has keys loaded...
Is there any way to implement the dedup semantic wherein it forces inheritance of the property even if the source didn't deduplicate? This is kind of a serious problem, as i'm still having to build new pools on dm-crypt since we can't do full replicas from existing pools into the ones if i were to utilize ZFS crypto on the receiving side.

tcaputi commented:
@sempervictus 
> still running into the problem of trying to send from unencrypted datasets to encrypted trees via zfs send -Rve - complains that it can't recv the unencrypted property. Recv side propertly (-o -x) not helping here. The target has keys loaded...

This will be fixed in a future PR. I actually have the code written, but it relies on these fixes first. I will make that PR as soon as this one is done. 

When you do `zfs send -R` you are sending the `encryption=off` property along with your data. The new patch will allow you to do `zfs recv -o encryption=...`, overriding this. Until then, unfortunately, you have to use a regular (not `-p` or `-R`) send to backup your data to an encrypted dataset. I'm really hoping to have this done soon, but the issues in #6639 are proving to be difficult to track down.

My apologies for the wait.

sempervictus commented:
Took a peek at #6639, that doesn't look fun, condolences and thanks for working on it.
Id ask for a way to do -o/-x now, but reading that, i dont think i _want_ to send encrypted datasets anywhere now :). We do have fully encrypted DS' at the datacenter which are sent every hour to backup nodes, but we keep them up with keys loaded since we actually access the backups for other reasons, so the -wp thing doesn't apply to that use case.
Thanks as always for digging through this, the implementation has been a Herculean effort, and i'm sorry to keep bugging you about the rough edges left over (when you're unfing kernel panics). Will keep an eye on this PR and that issue over the coming days.

behlendorf commented on include/sys/dmu.h:
```diff
@@ -797,9 +800,11 @@ int dmu_write_uio_dnode(dnode_t *dn, struct uio *uio, uint64_t size,
+    struct arc_buf *buf, dmu_tx_t *tx);
+void dmu_assign_arcbuf_by_dbuf(dmu_buf_t *handle, uint64_t offset,
+    struct arc_buf *buf, dmu_tx_t *tx);
+void dmu_assign_arcbuf_by_dbuf_impl(dmu_buf_t *handle, struct arc_buf *buf,
```
This is a good time to remove this `_impl` function from both the header and it implementations from `module/zfs/dmu.c`.  The single call site in `module/zfs/dmu_send.c` can be updated to do the cast as a simplification.

behlendorf commented on include/sys/dmu.h:
```diff
@@ -797,9 +800,11 @@ int dmu_write_uio_dnode(dnode_t *dn, struct uio *uio, uint64_t size,
-    dmu_tx_t *tx);
-void dmu_assign_arcbuf_impl(dmu_buf_t *handle, struct arc_buf *buf,
+void dmu_assign_arcbuf(objset_t *os, uint64_t object, uint64_t offset,
+    struct arc_buf *buf, dmu_tx_t *tx);
```
In order to avoid needlessly breaking the Lustre build, which depends on this symbol, let's do the following. 

* Rename `dmu_assign_arcbuf()` to `dmu_assign_arcbuf_by_dbuf()` as you've already done.  This is a better name and consistent with the other `_by_dnode/_by_dbuf` changes.
* Add a #define in the header to map the `dmu_assign_arcbuf` function to `dmu_assign_arcbuf_by_dbuf`.
* Update the one call site where you needed the new `dmu_assign_arcbuf()` to take the dnode hold itself so you don't need this wrapper.
* Concurrently let's give @adilger and @bzzz77 a heads up so we they can migrate to the new `_by_dbuf` interface.
* At some point in the future we can then retire the compatibility code and add the alternate version of `dmu_assign_arcbuf` proposed here.

tcaputi commented:
I made a small typo last push which caused some crashes during zfs-stress testing (which has now been corrected). As of now, the only problems now are as follows:

- The `CentOS 6 x86_64 (TEST)` worker fails to run `zpool freeze` in a few different tests for some reason. I have no explanations as to how encryption could be causing this on this worker alone.
- The `Ubuntu 14.04 i686 (TEST)` worker doesn't correctly recognize disks in use. At this point, I am fairly certain that this is a problem in that particular version of libblkid.

Aside from that, I believe all panics, crashes, error messages, and documentation issues have been taken care of. `ztest` is now also testing encrypted datasets which should be good for maintainability going forward. We're getting pretty close now (knock on wood) so I'm going to remove the WIP tag.

lundman commented:
Looks like the `lsize` variable was moved up some 20 lines, perhaps the second one is no longer needed?
https://github.com/datto/zfs/blob/94039fd8993babc6508b1d63bc3b76ada48e9d7a/module/zfs/dbuf.c#L3470

tcaputi commented:
Patch has been rebased on master and to match @behlendorf 's style recommendations.

github-usr-name commented:
Sorry if I'm being thick, and thanks for all of your work, but this is still failing in my case:
`pete@xxx ~ % zfs send -R zroot/mail@znap_2018-01-28-1200_hourly|zfs recv -x encryption -x keylocation -vdu -n zroot/cryptroot`
`cannot receive: invalid property 'encryption'`
`pete@xxx ~ % pacman -Qo zfs`
`/usr/bin/zfs is owned by zfs-utils-common-git 2018.01.18.r3248.7da8f8d81-1`
Anything obvious I've got wrong?

tcaputi commented:
@github-usr-name 
`zfs recv -x encryption` and `zfs recv -x keylocation` aren't implemented yet. Sorry. That's next on my list of things to work on.

github-usr-name commented:
Thanks for the reply. Not a problem, just good to know I've not screwed anything up w/ my mental model. Those encryption properties are awkward semantic exceptions to the inheritance rules.

On 28 January 2018 14:41:27 GMT+00:00, Tom Caputi <notifications@github.com> wrote:
>@github-usr-name 
>`zfs recv -x encryption` and `zfs recv -x keylocation` aren't
>implemented yet. Sorry. That's next on my list of things to work on.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/zfsonlinux/zfs/pull/6595#issuecomment-361067925

ajs124 commented:
@tcaputi any news on the `zfs recv -x encryption` and `zfs recv -x keylocation` support?

tcaputi commented:
It is next on my list once my currently outstanding PRs and issues are done. I should be able to get to it within the next couple weeks

Redsandro commented:
@tcaputi my apologies if I have missed this in different channels.

Can you tell us what tagged version includes the final encryption fixes and support for `zfs send test/old_broken_crypt | zfs receive test/new_crypt`?

I'm running my own compile with the old encryption. Was hoping the new Ubuntu Bionic 18.04 release [would inlcude this](https://launchpad.net/ubuntu/+source/zfs-linux/0.7.5-1ubuntu16.3) so I could just install the package, no questions asked, but I'm guessing it's too old (0.7.5). 0.7.8 is the first post-march release according to the zfsonlinux releases page.

adilger commented:
@Redsandro there is no encryption support in ZoL 0.7.x, only in git master (which will become ZoL 0.8.0).

Redsandro commented:
Thanks for clearing that up for me @adilger.

_*Redsandro writes down ZoL 0.8.0 in his agenda._

Redsandro commented:
Now testing ZFS `0.8.0-rc1`.

@adilger @tcaputi I remember the proper way to migrate pre-release encrypted datasets to new encrypted datasets was documented somewhere. However, looking through my issues and mentions, I cannot find where it was discussed. 

Can you point me to the docs? Or is it literally as simple as `zfs send pool/untagged_alpha_encryption | zfs receive test/tagged_080rc1_encryption`?

Would I have to create the new dataset with encryption first?

(Just to be sure. Not messing up a lengthy operation is preferable.)

tcaputi commented:
@Redsandro When you import the pool on the new version if you are effected you will see a message in `zpool status` with a brief description of how to send the data. There will also be a URL to some documentation with more verbose instructions (I can't remember what it is right now).

Redsandro commented:
@tcaputi @behlendorf I was confident enough to set things in motion. So far I'm happy that the old datasets can be mounted as read-only so we can copy them on the same pool. Really nice. :+1: 

There was however no description on _how_ to send the data, nor a URL to some documentation.

> __Suggestion:__ Put (shortened) url from your comment into the _status_ or _action_ text of `zpool status`

I will now write down everything I did after first reading the zpool description and then the linked documentation. Initially, all this achieves is making you think I am an idiot. But once you look past that, it might show where the brief zpool description or the documentation you linked can be improved by observing the mistakes I made after reading them.

On the one hand, this being a pre-release, why not make some improvements to the descriptions. On the other hand, I am not sure it's worth the effort to improve the documentation for something that only a select few are going to need/read.

```
$ sudo zpool import
   pool: zfs
  state: ONLINE
 status: Some supported features are not enabled on the pool.
 action: The pool can be imported using its name or numeric identifier, though
	some features will not be available without an explicit 'zpool upgrade'.
```

This sounds like I need to enter `zpool upgrade` first, if I want the new features.

```
$ sudo zpool upgrade

Every feature flags pool has all supported features enabled.
```

Okay, nevermind, I guess I need to enter this _after_ the import. The wording could be changed to emphasize that.

> __Suggestion:__ "... some features will not be available _until_ an explicit 'zpool upgrade' is ordered."

```
$ sudo zpool import zfs pool
$ sudo zpool status
  pool: pool
 state: ONLINE
status: Errata #3 detected.
	Existing encrypted datasets contain an on-disk incompatibility
	which needs to be corrected.
action: To correct the issue backup existing encrypted datasets to new
	encrypted datasets and destroy the old ones. 'zfs mount -o ro' can
	be used to temporarily mount existing encrypted datasets readonly.
```

There is no description on _how_ to send the data, nor a URL to some documentation. However, the documentation I found in your edited comment eventually helped me show the way. I am grateful that you built this possibility. It saves an entire in-between datadump for the datasets we used to test this new feature on.  :+1:

Let's proceed.

```
$ sudo zfs mount -l -o ro pool/data
$ sudo zfs create -o encryption=on -o keyformat=passphrase -o mountpoint=/mnt/data_new pool/data_new
$ sudo zfs send pool/data | pv | sudo zfs receive pool/data_new
warning: cannot send 'pool/data': target is busy; if a filesystem, it must not be mounted

$ sudo zfs unmount pool/data
$ sudo zfs unmount pool/data_new
$ sudo zfs send pool/data | pv | sudo zfs receive pool/data_new
cannot receive new filesystem stream: destination 'pool/data_new' exists

$ sudo zfs send pool/data | pv | sudo zfs receive pool/data_new -F
cannot receive new filesystem stream: zfs receive -F cannot be used to destroy an encrypted filesystem or overwrite an unencrypted one with an encrypted one

$ sudo zfs send pool/data | pv | sudo zfs receive pool/data_new/data_old
31.2GiB 0:12:36 [69.5MiB/s] ## It's working!
```

I never used "zfv receive" on what seems to be a subdirectory - didn't know this was legal - so I am not sure what is happening, but I assume it's exactly that: A subdirectory, and I can move the files up one level after the transfer is finished.

> __Suggestion:__ Mention in _"ZFS Errata \#\3"_ that the zfs send | zfs receive method requires both datasets to be unmounted, which is pretty obvious logically, but (1) I was tempted to mount the old dataset as advised in the `zpool status` message, and (2) the new dataset in the `zfs create` step from _"ZFS Errata \#\3"_ is automatically mounted, but the _"ZFS Errata \#\3"_ doesn't document an `unmount` step, so I didn't perform one either.

Redsandro commented:
> I assume it's exactly that: A subdirectory, and I can move the files up one level after the transfer is finished.

I was wrong. It's a separate dataset. Is moving data to the parent (same encryption options) a trivial or expensive operation?

tcaputi commented:
@Redsandro i dont see any reason why zfs send / recv should be slower than rsync. On the contrary, if you are sending a series of snapshots zfs send should (theoretically) always be faster

Redsandro commented:
@tcaputi send|recv is slower only if _"one is adamant on having the files in the encryption root"_, because you would have to `mv` the files afterwards, which is an expensive operation.

But perhaps I'm using zfs 'wrong', if there's such a thing. Perhaps the 'best practice' is to always use datasets inside a root (e.g. `pool/root/dataset`) as opposed to my setup (`pool/dataset`) which is expected by custom scripts and links.

I'm experimenting with this on a different machine. So far I found that children to a root (`pool/root/child`) are automatically mounted inside the root mountpoint, but then the root will not mount anymore, because its mountpoint is not empty.
